src/lua/mod.rs

```diff
@@ -218,61 +202,173 @@ fn thread_handle_message(sender: &Sender<LuaResponse>,
+            }
+            AnyLuaValue::LuaArray(lua_table)
+        }
+    }
+}
 
-        LuaQuery::SetValue { name: _name, val: _val } => {
-            panic!("thread: unimplemented LuaQuery::SetValue!");
-        },
+/// Converts an `AnyLuaValue` to a `Json`.
+///
+/// For an already-matched `LuaArray`, use `lua_array_to_json`.
+///
+/// For a `LuaArray` that should be mapped to a `JsonObject`,
+/// use `lua_object_to_json`.
+pub fn lua_to_json(lua: AnyLuaValue) -> Result<Json, ()> {
+    match lua {
+        AnyLuaValue::LuaNil => Ok(Json::Null),
+        AnyLuaValue::LuaString(val) => Ok(Json::String(val)),
+        AnyLuaValue::LuaNumber(val) => Ok(Json::F64(val)),
+        AnyLuaValue::LuaBoolean(val) => Ok(Json::Boolean(val)),
+        AnyLuaValue::LuaArray(arr) => lua_array_to_json(arr),
+        AnyLuaValue::LuaOther => Err(())
+    }
+}
 
-        LuaQuery::EmptyArray(_name) => {
-            panic!("thread: unimplemented LuaQuery::EmptyArray!");
-        },
+pub fn lua_array_to_json(arr: Vec<(AnyLuaValue, AnyLuaValue)>)
+                         -> Result<Json, ()> {
+    // Check if every key is a number
+    let mut counter = 0.0; // Account for first index?
 
-        _ => {
-            panic!("Unimplemented send type for lua thread!");
+    for &(ref key, ref _val) in &arr {
+        match *key {
+            AnyLuaValue::LuaNumber(num) => {
+                counter += num;
+            }
+            AnyLuaValue::LuaString(_) => {
+                break;
+            }
+            // Non-string keys are not allowed
+            _ => {
+                return Err(());
+            }
         }
     }
+
+    // Gauss' trick
+    let desired_sum = ((arr.len()) * (arr.len() + 1)) / 2;
+    if counter != desired_sum as f64 {
+        return lua_object_to_json(arr);
+    }
+
+    let mut json_arr: Vec<Json> = Vec::with_capacity(arr.len());
+
+    for (_key, val) in arr.into_iter() {
+        let lua_val = try!(lua_to_json(val));
+        json_arr.push(lua_val);
+    }
+    Ok(Json::Array(json_arr))
 }
 
-fn thread_send(sender: &Sender<LuaResponse>, response: LuaResponse) {
-    trace!("Called thread_send");
-    match sender.send(response) {
-        Err(_) => {
-            error!("thread: Unable to broadcast response!");
-            error!("thread: Shutting down in response to inability \
-                    to continue!");
-            panic!("Lua thread unable to communicate with main thread, \
-                    shutting down!");
+pub fn lua_object_to_json(obj: Vec<(AnyLuaValue, AnyLuaValue)>)
+                          -> Result<Json, ()> {
+    let mut json_obj: BTreeMap<String, Json> = BTreeMap::new();
+
+    for (key, val) in obj.into_iter() {
+        match key {
+            AnyLuaValue::LuaString(text) => {
+                json_obj.insert(text, try!(lua_to_json(val)));
+            },
+            AnyLuaValue::LuaNumber(ix) => {
+                json_obj.insert(format!("{}", ix), try!(lua_to_json(val)));
```

Comment:
Not using `ix.to_string`? I'm not sure if there is any additional cost to `format!`, but it might be easier to read/understand.